PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/ImplementationTypeCheck.java

```diff
@@ -0,0 +1,143 @@
+         *
+         * @param holder - the ProblemsHolder object to register the problem
+         */
+        ImplementationTypeVisitor(ProblemsHolder holder) {
+            this.holder = holder;
+        }
+
+        // Define constants for string literals
+        private static final RuleConfig RULE_CONFIG;
+        private static final boolean SKIP_WHOLE_RULE;
+        private static final String RULE_NAME = "ImplementationTypeCheck";
+
+        static {
+            RuleConfigLoader centralRuleConfigLoader = RuleConfigLoader.getInstance();
+
+            // Get the RuleConfig object for the rule
+            RULE_CONFIG = centralRuleConfigLoader.getRuleConfig(RULE_NAME);
+            SKIP_WHOLE_RULE = RULE_CONFIG.skipRuleCheck() || RULE_CONFIG.getAntiPatternMessageMap().isEmpty() || RULE_CONFIG.getListedItemsToCheck().isEmpty();
+        }
+
+        /**
+         * This method visits the variables in the code.
+         * It checks if the variable type is an Azure implementation type and flags it as a problem if it is.
+         */
+        @Override
+        public void visitVariable(@NotNull PsiVariable variable) {
+            super.visitVariable(variable);
+
+            if (SKIP_WHOLE_RULE) {
+                return;
+            }
+
+            // Get the type of the variable - This is the type of the variable declaration
+            // eg. List<String> myList = new ArrayList<>(); -> type = List<String>
+            PsiType type = variable.getType();
+
+            // Check if the type directly used is an implementation type
+            if (isImplementationType(type) && variable.getNameIdentifier() != null) {
+                holder.registerProblem(variable.getNameIdentifier(), RULE_CONFIG.getAntiPatternMessageMap().get("antiPatternMessage"));
+
+                // Check if the type extends or implements an implementation type
+            } else if (ExtendsOrImplementsImplementationType(type) && variable.getNameIdentifier() != null) {
+                holder.registerProblem(variable.getNameIdentifier(), RULE_CONFIG.getAntiPatternMessageMap().get("antiPatternMessage"));
+            }
+        }
+
+        /**
+         * This method checks if the type is a class from the Azure package and if it is an implementation type.
+         * It returns true if the type is an implementation type and false otherwise.
+         */
+        private boolean isImplementationType(PsiType type) {
+
+            if (!(type instanceof PsiClassType)) {
+                return false;
+            }
+            PsiClass psiClass = ((PsiClassType) type).resolve();
+
+            if (psiClass == null) {
+                return false;
+            }
+            // Check if the class is in the Azure package and if it is an implementation type
+            return psiClass.getQualifiedName().startsWith(RuleConfig.AZURE_PACKAGE_NAME) && psiClass.getQualifiedName().contains(RULE_CONFIG.getListedItemsToCheck().get(0));
```

Comment:
We should not use `get(0)` as this will only check the first element in the list of configured rules.

Reply:
There's only one entry and its in String form.

 We ended up with .get(0) for cases where there's only one String listedItemToCheck, but other rules have a list listedItemToCheck so instead of having a RULE_CONFIG attribute for String ItemToCheck and List listedItemToCheck we can just keep the list attribute, and when a String is encountered, it returns a list with one entry